### PR TITLE
Optimize form updates with entities and media files

### LINKED ISSUE #6807: Speed up form updates with 1000s of attached media files
~~Blocked by https://github.com/getodk/collect/pull/6806~~

A partner has identified several opportunities to speed up form updates with attached media files:

- [x] https://github.com/getodk/collect/pull/6803
- [x] [Optimize search for existing files](https://github.com/getodk/collect/pull/6808) (`ServerFormUseCases#searchForExistingMediaFile`)
- [ ] Consider not copying files to cache directory and back again when the form version is the same and the media files haven't changed

They have a target of "get the form refresh down to <5 seconds with 2500 media files" on their devices. Files are about 100-150 kB.

About copying files to cache, they say "We think removing the unnecessary file copy & hashing operations will cut >75% off the remaining time and avoid wear & tear on the device’s flash memory."

This is time-sensitive for them and we want to have it ready for a pilot round by July 21st.

### PR DESCRIPTION
**Pull request by seadowg**
Closes #6807

#### Why is this the best possible solution? Were any other approaches considered?

There's absolutely a better route here, but this should allow a fork to set `FeatureFlags.FASTER_FORM_UPDATES` to `true` to enable faster form updates with many media files and entity lists. 

The core change here is to detect the case where the form hasn't changed, but media files have and then skip downloading/copying media files. We can do that when using Central because we can make the assumption that if media files have changed and the OpenRosa form hash hasn't, then we'll just be downloading new media files and copying them to an existing form directory rather than creating a new one. This might also work for other servers, but we'll need to do additional manual and automated testing to verify. I also think we'd want to try and simplify our download logic more to handle these kinds of cases more explicitly: right now downloading a form (`ServerFormDownloader`) involves a series of steps that are pretty unrelated - we should break them into parts and have explicit branching for different scenarios (downloading a new form, updating a form, updating just media files etc).

I also changed our optimization around copying media files from previous form versions: it now only checks the last version or the "current" one in the case of just updating media files. This will probably be faster in most cases (as we're not scanning every previous version), but the main goal was to reduce the number of possible situations I had to consider.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

It should provide faster refreshes for forms with entity lists and many media files in both cases where the entity list is the same and when it changes (but not when the form version is updated or for initial download). There are definitely a bunch of risks here so checking form download and updates in general is a must, although we only need to check against Central initially (we'll see how broken this is with other servers later).

Another thing to point out is that we don't currently have an automated benchmark for actual form version updates for forms with many media files. These should remain as fast as they are in the current store version (v2025.2.0).

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by seadowg** [comment]
@tobiasmcnulty are you able to have a quick test with https://output.circle-artifacts.com/output/job/450371fe-a462-489b-b16e-d58c8f48a2b3/artifacts/0/selfSignedRelease.apk? I think that might get you another jump in speed.

**Comment by tobiasmcnulty** [comment]
Timings with 2500 ~100kB files:
- Initial media download: ~8 minutes
- Refresh with no server side changes (fresh install, only one local form version): 11 seconds (about 1-2 seconds faster than the last PR, I'm surprised the media file copy routine change didn't result in a greater improvement)
- Loading state after tapping "Start new form": 8 seconds

Test device:
- CPU: Unisoc T606 8-Core
- GPU: Arm Mali-G57 MP1 650MHz
- RAM: 8 GB
- Storage: 128GB
- Screen: 10.51-inch, 1200×1920 IPS
- OS Android 13

**Comment by seadowg** [comment]
@tobiasmcnulty thanks! I saw much larger improvements on a different device, so just wanted to check before getting more invasive.

**Comment by tobiasmcnulty** [comment]
@seadowg Regarding 808c16b, I noticed `fileResult.isNew` is based on the hash of the form XML, so it could provide a trigger for skipping form-specific code when the XML hasn't changed, if needed. I don't believe the form-related code is a bottleneck in our case though.

**Comment by seadowg** [comment]
> @seadowg Regarding https://github.com/getodk/collect/commit/808c16b77858de3009319b8659346e51ada92984, I noticed fileResult.isNew is based on the hash of the form XML, so it could provide a trigger for skipping form-specific code when the XML hasn't changed, if needed. I don't believe the form-related code is a bottleneck in our case though.

@tobiasmcnulty Yup that's right. If you look back through the commits you'll see that was an initial spike on involving skipping media file copies. I'd initially wanted a separate path for this, but it ended up feeling more risky than not doing that. You should be able to test with the APK for [808c16b](https://github.com/getodk/collect/pull/6811/commits/808c16b77858de3009319b8659346e51ada92984) once it's built.

File: collect_app/src/main/java/org/odk/collect/android/application/FeatureFlags.kt [hunk 1]
```diff
@@ -3,4 +3,5 @@ package org.odk.collect.android.application
 object FeatureFlags {
 
     const val NO_THEME_SETTING = true
+    const val FASTER_FORM_UPDATES = true
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/application/FeatureFlags.kt]: @grzesiek2010 we should change this back to `false` before merging!

Comment (tobiasmcnulty) [reply collect_app/src/main/java/org/odk/collect/android/application/FeatureFlags.kt]: What are the perceived risks with skipping the file copy, out of curiosity? Just so we know what to look out for.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/FeatureFlags.kt]: In non-Central servers you can do things like update the form name without updating the hash/version. That case alone introduces some scary possibilities due to download code using form name for file paths in some cases. I'm not 100% sure that that case would even break with this new optimization, but it's also not something I feel I'd be able to get confident about quickly and that's just one possible edge case with more permissive servers. In general the risk area is where the "update" doesn't fit our regular model (form version update or media file update).

Basically to get this kind of into "prod" we'd need to write a bunch more tests, test manually against servers like Kobo and Ona and also have it out as a beta before proper release. I'd like to get the server download code to a place where making these kinds of changes is much easier as well.

**Comment by tobiasmcnulty** [comment]
New timings: 🎆 🥇 
- Refresh with no server side changes: 4.5 seconds
- Loading state after tapping "Start new form": 3 seconds

Nice work! Skipping the copies is a big improvement. I haven't regression tested this or tested other devices yet, but on my device, in the Device Explorer, it looks like the media files are still where they should be, and they have the timestamps from when they were initially downloaded instead of when the form list was refreshed.

Unfortunately I feel like the loading state after tapping "Start new form" (#6810) is still an issue. I'll follow up there.

**Comment by seadowg** [comment]
> Unfortunately I feel like the loading state after tapping "Start new form" (https://github.com/getodk/collect/issues/6810) is still an issue. I'll follow up there.

Yeah, I wouldn't expect the changes here to make any difference. It's just focused on server updates.

**Comment by tobiasmcnulty** [comment]
> > Unfortunately I feel like the loading state after tapping "Start new form" (#6810) is still an issue. I'll follow up there.
> 
> Yeah, I wouldn't expect the changes here to make any difference. It's just focused on server updates.

As [noted here](https://github.com/getodk/collect/issues/6810#issuecomment-3079173632), the form update code is definitely running when one taps "Start new form," and thereby shaved off about 5 seconds (!) from that process via these updates.

File: collect_app/src/main/java/org/odk/collect/android/application/FeatureFlags.kt (hunk 1 above)
Comment (tobiasmcnulty) [reply collect_app/src/main/java/org/odk/collect/android/application/FeatureFlags.kt]: Makes sense, thanks!

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt [hunk 2]
```diff
@@ -203,13 +209,18 @@ object ServerFormUseCases {
...
     private fun searchForExistingMediaFile(
-        allFormVersionsSorted: List<Form>,
+        currentOrLastFormVersion: Form?,
         mediaFile: MediaFile
     ): File? {
-        return allFormVersionsSorted.map { form: Form ->
-            File(form.formMediaPath, mediaFile.filename)
-        }.firstOrNull { file: File ->
-            file.exists()
+        if (currentOrLastFormVersion != null) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt]: There could be une return:
```
        return if (currentOrLastFormVersion != null) {
            val candidate = File(currentOrLastFormVersion.formMediaPath, mediaFile.filename)
            if (candidate.exists()) {
                candidate
            } else {
                null
            }
        } else {
            null
        }
```

File: collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt [hunk 3]
```diff
@@ -63,7 +64,48 @@ class FormsUpdateBenchmarkTest {
...
+                it
+                    .clickGetSelected()
+                    .clickOKOnDialog(MainMenuPage())
+            }
+
+        benchmarker.assertResults()
+    }
+
+    /**
+     * Benchmarks the performance of updating forms. [THOUSAND_MEDIA_FILE_PROJECT_URL] should
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt]: Thsi should be `THOUSAND_MEDIA_FILE_ENTITY_LIST_PROJECT_URL`.

File: collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt [hunk 4]
```diff
@@ -63,7 +64,48 @@ class FormsUpdateBenchmarkTest {
...
+            .addProject()
+
+            // Download all forms
+            .clickGetBlankForm()
+            .clickGetSelected()
+            .clickOKOnDialog(MainMenuPage())
+
+            .clickGetBlankForm()
+            .benchmark(
+                "Redownloading a form with 1k media files and entity list when there are no updates",
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt:107]: Hmm if it is redownloading (with no updates) after opening the list of forms, the form should not be selected, so don't you need to click `select all` first?

File: README.md [hunk 5]
```diff
@@ -260,6 +260,7 @@ Maintainers keep a folder with a clean checkout of the code and use [jenv.be](ht
 - run `./gradlew releaseCheck`. If successful, a signed release will be at `collect_app/build/outputs/apk` (with an old version name)
 - verify a basic "happy path": scan a QR code to configure a new project, get a blank form, fill it, open the form map (confirms that the Google Maps key is correct), send form
 - run `./benchmark.sh` with a real device connected to verify performance
+  - To run benchmarks a project will need to be setup in Central with the benchmark forms and app users. The forms and entities needed for that are available [here](https://drive.google.com/drive/folders/1dPLvDY0LhVX-5qTUEs6EDoraDnLpUS0g?usp=drive_link).
```
Comment (grzesiek2010) [inline README.md]: Isn't `setup` a noun only?

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/download/ServerFormDownloader.java [hunk 6]
```diff
@@ -75,10 +75,12 @@ public void downloadForm(ServerFormDetails form, @Nullable ProgressReporter prog
 
         File tempDir = new File(cacheDir, "download-" + UUID.randomUUID().toString());
         tempDir.mkdirs();
+        String tempMediaPath = new File(tempDir, "media").getAbsolutePath();
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formmanagement/download/ServerFormDownloader.java]: Why do we need the changes here? I think you added something and then reverted that change, but not entirely.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/download/ServerFormDownloader.java (hunk 6 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/download/ServerFormDownloader.java]: It was just a readability change at the time - do you think it needs reverting?

File: collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt (hunk 4 above)
Comment (seadowg) [reply collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt:107]: No. Forms with updates are already selected.

File: README.md (hunk 5 above)
Comment (seadowg) [reply README.md]: Hah I didn't know that: https://www.merriam-webster.com/dictionary/setup. I think generally people use them interchangeably in my experience, but I'll change it here!

File: collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt (hunk 4 above)
Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt:107]: **Forms with updates are already selected** that's true but here it says  **when there are no updates**.

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/download/ServerFormDownloader.java (hunk 6 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/download/ServerFormDownloader.java]: If the rest was reverted and this part does not improve anything, I think it should be reverted as well.

File: collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt (hunk 4 above)
Comment (seadowg) [reply collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt:107]: Entity forms always count as having updates from Collect's UI. If you run through the steps you'll see this does work. Should I add a comment?

Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt:107]: > Entity forms always count as having updates from Collect's UI.

Ahh could you remind me why it works like that?

Comment (seadowg) [reply collect_app/src/androidTest/java/org/odk/collect/android/benchmark/FormsUpdateBenchmarkTest.kt:107]: This was actually the root of the problem: entity forms always count as being updated when we check the form list as the manifest reports a hash that we can't calculate using the local file. This means that we always go through a redownload when performing a Match Exactly sync which led to the slow-down with media file copies. This all kinds of work by accident right now due to the entity list CSV not matching the local file MD5 - it's one of the many things I'd like to make explicit in our download logic.

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt [hunk 7]
```diff
@@ -135,12 +139,14 @@ object ServerFormUseCases {
...
             } else {
-                val existingFile = searchForExistingMediaFile(allFormVersionsSorted, mediaFile)
+                val existingFile = searchForExistingMediaFile(currentOrLastFormVersion, mediaFile)
                 if (existingFile != null) {
                     val existingFileHash = existingFile.getMd5Hash()
 
                     if (existingFileHash.contentEquals(mediaFile.hash)) {
-                        FileUtils.copyFile(existingFile, tempMediaFile)
+                        if (!formToDownload.mediaOnlyUpdate) {
+                            copyFileToDirectory(existingFile, tempMediaDir)
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: I understand that manual re-downloading (**without entities**) is outside the scope of this issue, but I just wanted to clarify that it won't be optimized in this case. This means that if there are no updates but you open the list of forms to download and manually download a form, every file will still be copied as before....

...but that makes me wonder why you didn't just pass and use here `isNewerFormVersionAvailable`? If it is false and the media files are equal then there is no need to copy anything no matter if it is a manual or automatic update, with or without entities.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt (hunk 7 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: You mean just expose `isNewerFormVersionAvailable` calculated in `ServerFormDetailsFetcher` as a property on `ServerFormDetails`? That would probably also work, but I'd have to spend some time thinking it through. 

The core problem was that `ServerFormDetails#isUpdated` is `true` if there's a new form version **or** there's new/changed media files. My goal was to give myself a flag to access in `ServerFormUseCases` (and originally `ServerFormDownloader`) that shows that we're in the "no new form version, but different media files case" so I've sort of ended up with the negation of just exposing `isNewerFormVersionAvailable`. I did initially just expose both `isNewerFormVersionAvailable` and `areNewerMediaFilesAvailable` on `ServerFormDetails` but that ended up needing a lot of subtle changes so settled on just trying to add one additional flag.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: > ServerFormDetails#isUpdated is true if there's a new form version or there's new/changed media files.

Correct, but `isNewerFormVersionAvailable` used to determine `ServerFormDetails#isUpdated` is only calculated based on the form hash, so if it is false, we know it is not a new form and media files will be downloaded to the same (existing) directory. That should be all we need to know? Am I missing something?

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: `isUpdated` can also be true if there's just new/changed media files. We don't know if it's true because of a form update or because of new media files once we get to `ServerFormUseCases#downloadMediaFiles`. It's like this when running background match exactly syncs:

1. Form ID is not on device - `ServerFormDownloader#downloadForm` is run
2. Form ID is on device and the hash has changed - `ServerFormDownloader#downloadForm` is run
3. Form ID is on device and the hash hasn't changed, but the manifest has changed - `ServerFormDownloader#downloadForm` is run
4. Form ID is on device and the hash hasn't changed - Do nothing

`isUpdated` is true in case 2 and 3, but we only want to skip copying media files from the previous (or "current") version in 3. I think in ideal world we'd expose these 4 (or potentially more states) directly to the download code, and it could take different paths based on which state we're in rather than changing behaviour at different points in one string of logic through inferring what state we're in using booleans.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: And just to clarify, when redownloading a non-entity form with no update we're in state 4 (and `isUpdated` is `false`), but in the background we'll never actually bother doing the download.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: I'm not talking about using `isUpdated` but the result of [isNewerFormVersionAvailable](https://github.com/getodk/collect/pull/6811/files#diff-4b80c7ffa064b73c1f934a83594a97835d8d2ce8eff0c91273f64369726c78bdR64) I understand that `isUpdated` is something more - (`isNewerFormVersionAvailable` || `areNewerMediaFilesAvailable`).

My question was why you can't pass the result of [isNewerFormVersionAvailable](https://github.com/getodk/collect/pull/6811/files#diff-4b80c7ffa064b73c1f934a83594a97835d8d2ce8eff0c91273f64369726c78bdR64) to `ServerFormDetails` and copy/skip copying media files just based on that flag?

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: Right. So basically you're suggesting that the line could be:

``` 
if (formToDownload.isUpdated && formToDownload.isNewerFormVersionAvailable) {
```

So only copy when we're doing an actual form update (state 2), but not in a media file only update (state 3) and we don't bother feature flagging. Is that right?

I **think** this would work, but we'd have to also consider servers where the name/version can change without the hash changing (which introduces more states to think about). For example, in theory the hash is the only thing that matters for determing the "version", but we'd need to be really sure that there are no bugs in the download code that could lead to forms changing directories etc if the name changes, and currently I'm not confident in that. From the PR description:

> This might also work for other servers, but we'll need to do additional manual and automated testing to verify.

The idea here was to limit the need for adding new tests and for testing against many servers, and to just keep myself focused!

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: > Right. So basically you're suggesting that the line could be:

`if (formToDownload.isUpdated && formToDownload.isNewerFormVersionAvailable) {`

not really... I wanted to pass `isNewerFormVersionAvailable` to https://github.com/getodk/collect/pull/6811/files#diff-8fc110d7b8e9b70cdd37288b72c7498adc251005dd94cf03711b92550827d473R147 and use there, just like that:

```
if (existingFileHash.contentEquals(mediaFile.hash)) {
        if (isNewerFormVersionAvailable) {
            copyFileToDirectory(existingFile, tempMediaDir)
        }
}
```

> I think this would work, but we'd have to also consider servers where the name/version can change without the hash changing (which introduces more states to think about)

so you are just worried that a form could be updated without updating the hash (which in fact is a server bug)? And that's why you don't want to use my simpler approach?
Is it possible to determine if it is central? If so, we could introduce this performance improvement just for central as we know it works with it.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: > so you are just worried that for could be updated without updating the hash (which in fact is a server bug)?

That and any other situations we might not have considered. Also, I'm not sure that sort of case is a "bug" really. Central is just far stricter than most other Open Rosa implementations, but I don't think [the spec](https://docs.getodk.org/openrosa-form-list/) mandates that hash and version have to change together, or that form name should be consistent between versions etc.

> And that's why you don't want to use my simpler approach?

The main reason here is that I'm out from tomorrow until Wednesday and I don't feel like I have time to think about it properly. I'm happy for a different solution to be pursued after merging this, or you could close this and work on an alternative, but I won't have the time to review or work on an alternative before Monday. 

And to be clear, I definitely agree that my solution is not ideal and can be improved and that's absolutely something I want to have happen as part of v2025.3. I'm currently thinking we should prioritize https://github.com/getodk/collect/issues/6517 to push some rewrites of the download code as a whole for example.

> Is it possible to determine if it is central? If so, we could introduce this performance improvement just for central as we know it works with it.

I don't think it's currently possible to reliably determine that we're talking to Central which is why I'd gone with the `FeatureFlag` route so it could just be enabled in the fork.

**Comment by srujner** [comment]
**Tested with Success**

Verified on device with Android 14

Verified cases: 
- Regression check on entities and forms with media;
- Downloading/updating Forms; 
- Multiple and single media files update;
- Entity list update only and entity list with media update;

Also we compared times in this PR vs Store and it looks good: 

### **1000 media files 4kB and 1000 entities (Pull Request)**

**Downloading from**
Galaxy M23 - 10min 16 sec
Galaxy M12 - 10 min 10 sec
Pixel 7a - 5 min 10 sec (staging)
Redmi 9T- 10min 30 sec

**Publishing same form without changes and refreshing**
Galaxy M23 - 8s, 7s
Galaxy M12 - 10s, 11s
Pixel 7a - 4s, 4.5s
Redmi 9t - 9s, 9s

**Adding entity (Enketo/ delete ent) and refreshing** (compare 2)
Galaxy M23 - 5s, 5s
Galaxy M12 - 5,5s, 4,5s
Pixel 7a -  2.5s, 3s
Redmi 9T - 6.5s, 5s

**Adding entity, changing media file and refreshing**
Galaxy M23 - 10s, 10s 
Galaxy M12 - 15s, 15s
Pixel 7a - 6.5s
Redmi 9T - 13.5s, 12s

**Refreshing list of forms without changing media or entities**
Galaxy M23 - 8s, 3s
Galaxy M12 - 4,5s, 3s
Pixel 7a - 1.5s
Redmi 9T - 3s, 2.5s


### **1000 media files 4kB and 1000 entities (STORE Version)**

**Downloading from**
Galaxy M23 - 10 min 15sec
Galaxy M12 10 min 16 sec
Pixel 7a - fail, fail, 5min
Redmi 9T - fail, 14min

**Publishing same form without changes and refreshing**
Galaxy M23 - 10s, 10s
Galaxy M12 11s, 12s
Pixel 7a - 4.5 s, 4.5s
Redmi 9T - 11s, 10s

**Adding entity and refreshing**
Galaxy M23 - 10s
Galaxy M12 - 15s, 15s
Pixel 7a - 4.5s, 3.7s
Redmi 9T - 17s, 12s

**Adding entity, changing media file and refreshing**
Galaxy M23 - 16s
Galaxy M12 - 24s
Pixel 7a- 8s

**Refreshing list of forms without changing media or entities**
Galaxy M23 - 7s, 6s
Galaxy M12 - 11s, 10s
Redmi 9T - 10s

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt (hunk 7 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: Ok we can merge this version if it is accepted by the QA team, but I feel like I'm having a hard time understanding that difference...
let me describe it:
1. We skip copying media files that already exist if `mediaOnlyUpdate` is true.
2. `mediaOnlyUpdate` is true if `!isNewerFormVersionAvailable && areNewerMediaFilesAvailable`.
3. That means we relay on that hash that might not be properly updated in this case as well.
4. If we just passed `!isNewerFormVersionAvailable` instead of `!isNewerFormVersionAvailable && areNewerMediaFilesAvailable` by skipping `areNewerMediaFilesAvailable` the logic would be simpler but the difference would be that we would skip copying media files that exist not only if there are some new media files but no matter if there are new or not.

Why would that be wrong? How does checking if there are new media files help you? I'm feeling kind of under the weather so sorry if my thinking is wrong :smile:

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: > If we just passed !isNewerFormVersionAvailable instead of !isNewerFormVersionAvailable && areNewerMediaFilesAvailable by skipping areNewerMediaFilesAvailable the logic would be simpler but the difference would be that we would skip copying media files that exist not only if there are some new media files but no matter if there are new or not.

I think (and hope) you're right here, but I'd need to actually see it in place and passing the tests and benchmarks, and we'd need to get Caktus to test again, which I don't have time to do and feel confident about. I think also we come back to this:

> That means we relay on that hash that might not be properly updated in this case as well.

So we'd still want to have the feature flag because we don't really have time to manually test against other servers and get confidence there aren't breaking edge cases. Without the feature flag it feels like a change we'd want going through beta rather than straight out as a hotfix. Given that, I think you end up with a similar level of complexity - it just ends up in `downloadMediaFiles` code rather than the `fetchFormDetails`. I think we should definitely try out your solution after though and see if it can work outside just the Central world as it's definitely nicer (especially once there isn't a feature flag) and it should also speed up the redownload case.

**Comment by WKobus** [comment]
**Tested with success**

Verified on device with Android 11

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt (hunk 7 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: >I think we should definitely try out your solution after though and see if it can work outside just the Central world as it's definitely nicer (especially once there isn't a feature flag) and it should also speed up the redownload case.

Ok, let's do that once you are back, maybe.
`areNewerMediaFilesAvailable` is always true for forms with entities as we discussed above, so for such forms, it doesn't matter if we use it or not, as in your logic `!isNewerFormVersionAvailable && areNewerMediaFilesAvailable` is the same as `!isNewerFormVersionAvailable && true`. 
There could be some difference in the case of forms without entities, but I don't think so.

**Comment by dbemke** [comment]
**Tested with success**

Verified on device with Android 10, 14

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt (hunk 7 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormUseCases.kt:148]: Sounds good! I think we're trying to do a two-week feature in 2–3 days here, so I think it's ok to end up in a situation where it only works in the one specific case we know needs it for now. I've got a note to re-discuss priorities next week based on the work we've done this week and at the end of last week on these optimizations.

**Review by grzesiek2010 — APPROVED** [review]
